File: test_rubiks_cube.py
"""
Test script for Rubik's Cube Solver
Tests various functionalities without using a testing framework
"""

from rubiks_cube import RubiksCube, SimpleSolver, reverse_moves


def test_initial_state():
    """Test that a new cube is in solved state"""
    print("Test 1: Initial State")
    cube = RubiksCube()
    assert cube.is_solved(), "New cube should be solved"
    print("✓ New cube is in solved state")
    print()


def test_single_move():
    """Test single move execution"""
    print("Test 2: Single Move Execution")
    cube = RubiksCube()

    # Test R move
    cube.move_R()
    assert not cube.is_solved(), "Cube should not be solved after R move"
    print("✓ R move executed, cube is scrambled")

    # Undo with Ri
    cube.move_Ri()
    assert cube.is_solved(), "Cube should be solved after R then Ri"
    print("✓ Ri move undoes R move correctly")
    print()


def test_all_basic_moves():
    """Test all basic moves and their inverses"""
    print("Test 3: All Basic Moves")
    moves = ['R', 'L', 'U', 'D', 'F', 'B']

    for move in moves:
        cube = RubiksCube()
        # Execute move
        getattr(cube, f'move_{move}')()
        assert not cube.is_solved(), f"Cube should not be solved after {move}"

        # Execute inverse
        getattr(cube, f'move_{move}i')()
        assert cube.is_solved(), f"Cube should be solved after {move} then {move}i"
        print(f"✓ {move} and {move}i moves work correctly")

    print()


def test_move_sequence():
    """Test executing a sequence of moves"""
    print("Test 4: Move Sequence Execution")
    cube = RubiksCube()

    # Execute a sequence
    sequence = "R U R' U'"
    cube.execute_moves(sequence)
    assert not cube.is_solved(), "Cube should be scrambled after sequence"
    print(f"✓ Executed sequence: {sequence}")

    # Execute inverse sequence
    cube.execute_moves("U Ri U Ri")
    assert cube.is_solved(), "Cube should be solved after inverse sequence"
    print("✓ Inverse sequence restored cube to solved state")
    print()


def test_commutator():
    """Test that certain move sequences return to solved state"""
    print("Test 5: Commutator Property")
    cube = RubiksCube()

    # A commutator: R U Ri Ui should affect only some pieces
    # Repeating it 6 times should return to solved state
    sequence = "R U Ri Ui"
    for i in range(6):
        cube.execute_moves(sequence)

    assert cube.is_solved(), "Cube should return to solved state after 6 commutators"
    print(f"✓ Executed '{sequence}' 6 times, cube returned to solved state")
    print()


def test_face_rotations():
    """Test that rotating a face 4 times returns it to original state"""
    print("Test 6: Face Rotation Property")
    moves = ['R', 'L', 'U', 'D', 'F', 'B']

    for move in moves:
        cube = RubiksCube()
        # Rotate 4 times
        for _ in range(4):
            getattr(cube, f'move_{move}')()

        assert cube.is_solved(), f"4 {move} moves should return cube to solved state"
        print(f"✓ Four {move} moves return to solved state")

    print()


def test_scramble_and_reverse():
    """Test scrambling and reversing moves"""
    print("Test 7: Scramble and Reverse")
    cube = RubiksCube()

    # Scramble with a sequence
    scramble = "R U R' U' L' U L U F F R' L F' R L'"
    cube.scramble(scramble)
    assert not cube.is_solved(), "Cube should be scrambled"
    print(f"✓ Cube scrambled with: {scramble}")

    # Reverse the scramble
    reverse = reverse_moves(scramble)
    cube.execute_moves(reverse)
    assert cube.is_solved(), "Cube should be solved after reversing scramble"
    print(f"✓ Reversed scramble: {' '.join(reverse)}")
    print("✓ Cube returned to solved state")
    print()


def test_cube_copy():
    """Test that cube copy works correctly"""
    print("Test 8: Cube Copy")
    cube1 = RubiksCube()
    cube1.execute_moves("R U R' U'")

    cube2 = cube1.copy()

    # Modify cube2
    cube2.move_F()

    # Check that cube1 is unchanged
    cube1_state = str(cube1)
    cube2_state = str(cube2)
    assert cube1_state != cube2_state, "Copied cube should be independent"
    print("✓ Cube copy is independent of original")
    print()


def test_is_solved():
    """Test the is_solved method"""
    print("Test 9: Is Solved Detection")

    # Solved cube
    cube = RubiksCube()
    assert cube.is_solved(), "New cube should be detected as solved"
    print("✓ Solved cube detected correctly")

    # Scrambled cube
    cube.execute_moves("R U F")
    assert not cube.is_solved(), "Scrambled cube should be detected as unsolved"
    print("✓ Unsolved cube detected correctly")
    print()


def test_complex_scramble():
    """Test with a more complex scramble"""
    print("Test 10: Complex Scramble")
    cube = RubiksCube()

    # A longer scramble
    scramble = "R L U D F B R' L' U' D' F' B' R U F R' U' F'"
    cube.scramble(scramble)
    print(f"✓ Applied complex scramble: {scramble}")

    # Verify it's scrambled
    assert not cube.is_solved(), "Cube should be scrambled"
    print("✓ Cube is in scrambled state")

    # Apply reverse
    reverse = reverse_moves(scramble)
    cube.execute_moves(reverse)
    assert cube.is_solved(), "Cube should be solved after reverse"
    print("✓ Complex scramble successfully reversed")
    print()


def test_move_equivalence():
    """Test that moving 3 times counterclockwise equals 1 clockwise"""
    print("Test 11: Move Equivalence")

    for move in ['R', 'L', 'U', 'D', 'F', 'B']:
        cube1 = RubiksCube()
        cube2 = RubiksCube()

        # One clockwise
        getattr(cube1, f'move_{move}')()

        # Three counterclockwise
        for _ in range(3):
            getattr(cube2, f'move_{move}i')()

        # States should be identical
        assert str(cube1) == str(cube2), f"{move} ≠ 3×{move}i"
        print(f"✓ {move} ≡ 3×{move}i")

    print()


def test_sexy_move():
    """Test the 'sexy move' (R U R' U') pattern"""
    print("Test 12: Sexy Move Pattern")
    cube = RubiksCube()

    # The sexy move repeated 6 times should return to solved
    sexy_move = "R U Ri Ui"
    for i in range(6):
        cube.execute_moves(sexy_move)

    assert cube.is_solved(), "Sexy move ×6 should solve cube"
    print(f"✓ Sexy move '{sexy_move}' repeated 6 times returns to solved state")
    print()


def test_solver_basic():
    """Test the basic solver"""
    print("Test 13: Solver - Already Solved")
    cube = RubiksCube()
    solver = SimpleSolver(cube)

    solution = solver.solve()
    assert cube.is_solved(), "Solver should maintain solved state"
    assert len(solution) == 0, "No moves needed for solved cube"
    print("✓ Solver recognizes already-solved cube")
    print()


def run_all_tests():
    """Run all tests"""
    print("=" * 60)
    print("RUBIK'S CUBE SOLVER - TEST SUITE")
    print("=" * 60)
    print()

    tests = [
        test_initial_state,
        test_single_move,
        test_all_basic_moves,
        test_move_sequence,
        test_commutator,
        test_face_rotations,
        test_scramble_and_reverse,
        test_cube_copy,
        test_is_solved,
        test_complex_scramble,
        test_move_equivalence,
        test_sexy_move,
        test_solver_basic,
    ]

    passed = 0
    failed = 0

    for test in tests:
        try:
            test()
            passed += 1
        except AssertionError as e:
            failed += 1
            print(f"✗ Test failed: {test.__name__}")
            print(f"  Error: {e}")
            print()
        except Exception as e:
            failed += 1
            print(f"✗ Test error: {test.__name__}")
            print(f"  Exception: {e}")
            print()

    print("=" * 60)
    print(f"TEST RESULTS: {passed} passed, {failed} failed")
    print("=" * 60)

    if failed == 0:
        print("\n🎉 All tests passed!")
    else:
        print(f"\n⚠️  {failed} test(s) failed")

    return failed == 0


def demo_cube():
    """Demonstrate cube functionality"""
    print("\n" + "=" * 60)
    print("DEMONSTRATION")
    print("=" * 60)
    print()

    print("Initial solved cube:")
    cube = RubiksCube()
    print(cube)
    print()

    print("After scramble 'R U R' U' F' L' U L U':")
    cube.scramble("R U Ri Ui Fi Li U L Ui")
    print(cube)
    print()

    print("Cube is solved:", cube.is_solved())
    print()


if __name__ == "__main__":
    # Run all tests
    success = run_all_tests()

    # Run demonstration
    demo_cube()

    # Exit with appropriate code
    exit(0 if success else 1)

File: test-rubiks-cube.js
/**
 * Testing Script for Rubik's Cube Solver
 * This script tests various functionalities without using a testing framework
 */

const RubiksCube = require('./rubiks-cube-solver');

// Test utilities
let testsPassed = 0;
let testsFailed = 0;
const failedTests = [];

function assert(condition, testName) {
  if (condition) {
    testsPassed++;
    console.log(`✓ ${testName}`);
  } else {
    testsFailed++;
    failedTests.push(testName);
    console.log(`✗ ${testName}`);
  }
}

function assertEqual(actual, expected, testName) {
  const condition = JSON.stringify(actual) === JSON.stringify(expected);
  assert(condition, testName);
  if (!condition) {
    console.log(`  Expected: ${JSON.stringify(expected)}`);
    console.log(`  Actual: ${JSON.stringify(actual)}`);
  }
}

function printTestSummary() {
  console.log('\n' + '='.repeat(50));
  console.log('TEST SUMMARY');
  console.log('='.repeat(50));
  console.log(`Total Tests: ${testsPassed + testsFailed}`);
  console.log(`Passed: ${testsPassed}`);
  console.log(`Failed: ${testsFailed}`);

  if (failedTests.length > 0) {
    console.log('\nFailed Tests:');
    failedTests.forEach(test => console.log(`  - ${test}`));
  }
  console.log('='.repeat(50) + '\n');
}

// Test 1: Cube Initialization
console.log('\n--- Test Suite: Cube Initialization ---\n');

function testCubeInitialization() {
  const cube = new RubiksCube();

  assert(cube.faces.U.every(sticker => sticker === 'W'),
    'Up face initialized with white');
  assert(cube.faces.D.every(sticker => sticker === 'Y'),
    'Down face initialized with yellow');
  assert(cube.faces.F.every(sticker => sticker === 'G'),
    'Front face initialized with green');
  assert(cube.faces.B.every(sticker => sticker === 'B'),
    'Back face initialized with blue');
  assert(cube.faces.L.every(sticker => sticker === 'O'),
    'Left face initialized with orange');
  assert(cube.faces.R.every(sticker => sticker === 'R'),
    'Right face initialized with red');
  assert(cube.isSolved(), 'Newly initialized cube is solved');
  assertEqual(cube.moveHistory.length, 0, 'Move history is empty on initialization');
}

testCubeInitialization();

// Test 2: Basic Moves
console.log('\n--- Test Suite: Basic Moves ---\n');

function testBasicMoves() {
  const cube = new RubiksCube();

  // Test R move
  cube.R();
  assert(!cube.isSolved(), 'Cube is not solved after R move');
  assertEqual(cube.moveHistory.length, 1, 'Move history contains 1 move after R');
  assertEqual(cube.moveHistory[0], 'R', 'Move history records R move correctly');

  // Test R' move (should undo R)
  cube.RPrime();
  assert(cube.isSolved(), "Cube is solved after R followed by R'");
  assertEqual(cube.moveHistory.length, 2, 'Move history contains 2 moves');

  // Test U move
  const cube2 = new RubiksCube();
  cube2.U();
  assert(!cube2.isSolved(), 'Cube is not solved after U move');

  // Test F move
  const cube3 = new RubiksCube();
  cube3.F();
  assert(!cube3.isSolved(), 'Cube is not solved after F move');
}

testBasicMoves();

// Test 3: Move Sequences
console.log('\n--- Test Suite: Move Sequences ---\n');

function testMoveSequences() {
  const cube = new RubiksCube();

  // Test simple sequence
  cube.executeSequence(['R', 'U', "R'", "U'"]);
  assert(cube.isSolved(), "Cube is solved after [R, U, R', U']");
  assertEqual(cube.moveHistory.length, 4, 'Move history contains 4 moves');

  // Test longer sequence
  const cube2 = new RubiksCube();
  cube2.executeSequence(['R', 'R', 'R', 'R']);
  assert(cube2.isSolved(), 'Four R moves return cube to solved state');

  // Test mixed sequence
  const cube3 = new RubiksCube();
  cube3.executeSequence(['U', 'U', 'U', 'U']);
  assert(cube3.isSolved(), 'Four U moves return cube to solved state');
}

testMoveSequences();

// Test 4: Cube Cloning
console.log('\n--- Test Suite: Cube Cloning ---\n');

function testCubeCloning() {
  const cube = new RubiksCube();
  cube.executeSequence(['R', 'U', 'F']);

  const clonedCube = cube.clone();

  assertEqual(clonedCube.faces.U, cube.faces.U, 'Cloned cube has same U face');
  assertEqual(clonedCube.faces.R, cube.faces.R, 'Cloned cube has same R face');
  assertEqual(clonedCube.moveHistory, cube.moveHistory, 'Cloned cube has same move history');

  // Modify original and ensure clone is unchanged
  cube.R();
  assert(cube.moveHistory.length !== clonedCube.moveHistory.length,
    'Cloned cube move history is independent');
}

testCubeCloning();

// Test 5: Scrambling
console.log('\n--- Test Suite: Scrambling ---\n');

function testScrambling() {
  const cube = new RubiksCube();
  const scrambleSequence = cube.scramble(10);

  assertEqual(scrambleSequence.length, 10, 'Scramble generates 10 moves');
  assertEqual(cube.moveHistory.length, 10, 'Move history contains scramble moves');

  // Most likely not solved after 10 random moves
  // (There's a tiny chance it could be, but extremely unlikely)
  const cube2 = new RubiksCube();
  cube2.scramble(20);
  // We can't assert it's not solved with 100% certainty, but check move history
  assertEqual(cube2.moveHistory.length, 20, 'Scramble with 20 moves records correctly');
}

testScrambling();

// Test 6: Solving
console.log('\n--- Test Suite: Solving ---\n');

function testSolving() {
  const cube = new RubiksCube();
  const scrambleSequence = cube.scramble(15);

  console.log(`  Scramble: ${scrambleSequence.join(' ')}`);

  const solution = cube.solve();
  console.log(`  Solution: ${solution.join(' ')}`);

  assert(solution.length > 0, 'Solver returns non-empty solution');
  assertEqual(solution.length, 15, 'Solution has same length as scramble');

  // Apply solution
  cube.executeSequence(solution);
  assert(cube.isSolved(), 'Cube is solved after applying solution');
}

testSolving();

// Test 7: Face Rotation
console.log('\n--- Test Suite: Face Rotation ---\n');

function testFaceRotation() {
  const cube = new RubiksCube();
  const originalFace = [...cube.faces.R];

  cube.rotateFaceClockwise('R');
  assert(JSON.stringify(cube.faces.R) !== JSON.stringify(originalFace),
    'Face changes after clockwise rotation');

  // Rotate 4 times should return to original
  cube.rotateFaceClockwise('R');
  cube.rotateFaceClockwise('R');
  cube.rotateFaceClockwise('R');
  assertEqual(cube.faces.R, originalFace,
    'Four clockwise rotations return face to original state');
}

testFaceRotation();

// Test 8: Move Inverses
console.log('\n--- Test Suite: Move Inverses ---\n');

function testMoveInverses() {
  // Test R and R'
  const cube1 = new RubiksCube();
  cube1.R();
  cube1.RPrime();
  assert(cube1.isSolved(), "R followed by R' solves cube");

  // Test U and U'
  const cube2 = new RubiksCube();
  cube2.U();
  cube2.UPrime();
  assert(cube2.isSolved(), "U followed by U' solves cube");

  // Test F and F'
  const cube3 = new RubiksCube();
  cube3.F();
  cube3.FPrime();
  assert(cube3.isSolved(), "F followed by F' solves cube");

  // Test reverse order
  const cube4 = new RubiksCube();
  cube4.RPrime();
  cube4.R();
  assert(cube4.isSolved(), "R' followed by R solves cube");
}

testMoveInverses();

// Test 9: Complex Patterns
console.log('\n--- Test Suite: Complex Patterns ---\n');

function testComplexPatterns() {
  // Test sexy move (R U R' U') repeated 6 times returns to solved
  const cube = new RubiksCube();
  for (let i = 0; i < 6; i++) {
    cube.executeSequence(['R', 'U', "R'", "U'"]);
  }
  assert(cube.isSolved(), "Sexy move repeated 6 times solves cube");

  // Test checkerboard pattern and back
  const cube2 = new RubiksCube();
  const pattern = ['R', 'R', 'U', 'U', 'F', 'F'];
  cube2.executeSequence(pattern);
  assert(!cube2.isSolved(), 'Checkerboard pattern is not solved state');
  cube2.executeSequence(pattern);
  assert(cube2.isSolved(), 'Checkerboard pattern repeated twice solves cube');
}

testComplexPatterns();

// Test 10: Edge Cases
console.log('\n--- Test Suite: Edge Cases ---\n');

function testEdgeCases() {
  // Empty sequence
  const cube1 = new RubiksCube();
  cube1.executeSequence([]);
  assert(cube1.isSolved(), 'Empty sequence keeps cube solved');

  // Single move repeated many times
  const cube2 = new RubiksCube();
  for (let i = 0; i < 100; i++) {
    cube2.R();
  }
  assert(cube2.isSolved(), '100 R moves (4 cycles of 25) returns to solved');

  // Test invalid move handling
  const cube3 = new RubiksCube();
  try {
    cube3.executeSequence(['X']);
    assert(false, 'Invalid move should throw error');
  } catch (error) {
    assert(true, 'Invalid move throws error correctly');
  }
}

testEdgeCases();

// Print final summary
printTestSummary();

// Exit with appropriate code
process.exit(testsFailed > 0 ? 1 : 0);
